File: library/src/main/java/com/grosner/dbflow/sql/migration/AlterTableMigration.java
package com.grosner.dbflow.sql.migration;

import android.database.sqlite.SQLiteDatabase;

import com.grosner.dbflow.config.FlowManager;
import com.grosner.dbflow.sql.QueryBuilder;
import com.grosner.dbflow.structure.Model;

import java.util.ArrayList;
import java.util.List;

/**
 * Author: andrewgrosner
 * Contributors: { }
 * Description: Provides a very nice way to alter a single table quickly and easily.
 */
public class AlterTableMigration<ModelClass extends Model> extends BaseMigration {

    private final Class<ModelClass> mTable;
    private QueryBuilder mQuery;
    private QueryBuilder mRenameQuery;
    private ArrayList<QueryBuilder> mColumnDefinitions;
    private String mOldTableName;

    public AlterTableMigration(Class<ModelClass> table) {
        mTable = table;
    }

    @Override
    public void onPreMigrate() {
        mQuery = new QueryBuilder().append("ALTER").appendSpaceSeparated("TABLE");
    }

    @Override
    public void migrate(SQLiteDatabase database) {
        // "ALTER TABLE "
        String sql = mQuery.getQuery();
        String tableName = FlowManager.getTableName(mTable);

        // "{oldName}  RENAME TO {newName}"
        // Since the structure has been updated already, the manager knows only the new name.
        if (mRenameQuery != null) {
            database.execSQL(sql + mOldTableName +
                    mRenameQuery.getQuery() + tableName);
        }

        // We have column definitions to add here
        // ADD COLUMN columnName {type}
        if (mColumnDefinitions != null) {
            sql = sql + tableName;
            for (QueryBuilder columnDefinition : mColumnDefinitions) {
                database.execSQL(sql + " ADD COLUMN " + columnDefinition.getQuery());
            }
        }
    }

    @Override
    public void onPostMigrate() {
        // cleanup and make fields eligible for garbage collection
        mQuery = null;
        mRenameQuery = null;
        mColumnDefinitions = null;
    }

    /**
     * Call this to rename a table to a new name, such as changing either the {@link com.grosner.dbflow.structure.Model} class name
     * or by changing the name through a {@link com.grosner.dbflow.structure.Table}
     *
     * @param oldName The new name to call the table.
     * @return
     */
    public AlterTableMigration<ModelClass> renameFrom(String oldName) {
        mOldTableName = oldName;
        mRenameQuery = new QueryBuilder().append(" RENAME").appendSpaceSeparated("TO");
        return this;
    }

    /**
     * Add a column to the DB. This does not necessarily need to be reflected in the {@link ModelClass},
     * but it is recommended.
     *
     * @param columnType
     * @param columnName
     * @return
     */
    public AlterTableMigration<ModelClass> addColumn(Class columnType, String columnName) {
        if (mColumnDefinitions == null) {
            mColumnDefinitions = new ArrayList<QueryBuilder>();
        }

        QueryBuilder queryBuilder = new QueryBuilder()
                .append(columnName).appendSpace().appendType(columnType.getName());
        mColumnDefinitions.add(queryBuilder);

        return this;
    }

    public String getRenameQuery() {
        QueryBuilder queryBuilder = new QueryBuilder(mQuery.getQuery()).append(mOldTableName)
                .append(mRenameQuery).append(FlowManager.getTableName(mTable));
        return queryBuilder.getQuery();
    }

    public List<String> getColumnDefinitions() {
        String sql = mQuery.getQuery() + FlowManager.getTableName(mTable);
        List<String> columnDefinitions = new ArrayList<String>();

        if (mColumnDefinitions != null) {
            for (QueryBuilder columnDefinition : mColumnDefinitions) {
                QueryBuilder queryBuilder = new QueryBuilder(sql).appendSpaceSeparated("ADD COLUMN").append(columnDefinition.getQuery());
                columnDefinitions.add(queryBuilder.getQuery());
            }
        }

        return columnDefinitions;
    }
}

File: lib/src/main/kotlin/com/dbflow5/migration/AlterTableMigration.kt
package com.dbflow5.migration

import androidx.annotation.CallSuper
import com.dbflow5.adapter.SQLObjectAdapter
import com.dbflow5.appendQuotedIfNeeded
import com.dbflow5.database.DatabaseWrapper
import com.dbflow5.query.select
import com.dbflow5.quoteIfNeeded
import com.dbflow5.sql.SQLiteType
import com.dbflow5.stripQuotes

/**
 * Description: Provides a very nice way to alter a single table quickly and easily.
 */
open class AlterTableMigration<T : Any>(
    /**
     * The table adapter to ALTER
     */
    adapterGetter: () -> SQLObjectAdapter<T>
) : BaseMigration() {

    protected val adapter by lazy(adapterGetter)

    /**
     * The query to rename the table with
     */
    private var renameQuery: String? = null

    /**
     * The columns to ALTER within a table
     */
    private val internalColumnDefinitions: MutableList<String> by lazy { mutableListOf() }

    private val columnNames: MutableList<String> by lazy { mutableListOf() }

    /**
     * The old name of the table before renaming it
     */
    private var oldTableName: String? = null

    override fun migrate(database: DatabaseWrapper) {
        // "ALTER TABLE "
        var sql = ALTER_TABLE
        val tableName = adapter.name

        // "{oldName}  RENAME TO {newName}"
        // Since the structure has been updated already, the manager knows only the new name.
        renameQuery?.let { renameQuery ->
            database.execSQL(buildString {
                append(sql)
                appendQuotedIfNeeded(oldTableName)
                append(renameQuery)
                append(tableName)
            })
        }

        // We have column definitions to add here
        // ADD COLUMN columnName {type}
        if (internalColumnDefinitions.isNotEmpty()) {
            (select from adapter limit 0).cursor(database)?.use { cursor ->
                sql = "$sql$tableName"
                for (i in internalColumnDefinitions.indices) {
                    val columnDefinition = internalColumnDefinitions[i]
                    val columnName = columnNames[i].stripQuotes()
                    if (cursor.getColumnIndex(columnName) == -1) {
                        database.execSQL("$sql ADD COLUMN $columnDefinition")
                    }
                }
            }
        }
    }

    @CallSuper
    override fun onPostMigrate() {
        // cleanup and make fields eligible for garbage collection
        renameQuery = null
        internalColumnDefinitions.clear()
        columnNames.clear()
    }

    /**
     * Call this to rename a table to a new name.
     *
     * @param oldName The new name to call the table.
     * @return This instance
     */
    fun renameFrom(oldName: String) = apply {
        oldTableName = oldName
        renameQuery = " RENAME TO "
    }

    /**
     * Add a column to the DB. This does not necessarily need to be reflected in the [T],
     * but it is recommended.
     *
     * @param sqLiteType The type of column represented in the DB.
     * @param columnName The name of the column to add. Use the "_Table" class for the specified table.
     * @param defaultValue the default value as a String representation.
     * Leaving parameter as null leaves the case out.
     * For NULL column add defaultValue = "NULL". Encapsulate the value in quotes "\'name\'" if it's a string.
     * @return This instance
     */
    @JvmOverloads
    fun addColumn(sqLiteType: SQLiteType, columnName: String, defaultValue: String? = null) =
        apply {
            var addStatement = "${columnName.quoteIfNeeded()} ${sqLiteType.name}"
            if (defaultValue != null) {
                addStatement += " DEFAULT $defaultValue"
            }
            internalColumnDefinitions.add(addStatement)
            columnNames.add(columnName)
        }

    /**
     * Add a column to the DB. This does not necessarily need to be reflected in the [T],
     * but it is recommended.
     *
     * @param sqLiteType      The type of column that pertains to an [SQLiteType]
     * @param columnName      The name of the column to add. Use the "$Table" class for the specified table.
     * @param referenceClause The clause of the references that this foreign key points to.
     * @return This instance
     */
    fun addForeignKeyColumn(sqLiteType: SQLiteType, columnName: String, referenceClause: String) =
        apply {
            internalColumnDefinitions.add("${columnName.quoteIfNeeded()} ${sqLiteType.name} REFERENCES $referenceClause")
            columnNames.add(columnName)
        }

    /**
     * @return The query that renames the table.
     */
    fun getRenameQuery(): String = buildString {
        append(ALTER_TABLE)
        appendQuotedIfNeeded(oldTableName)
        append(renameQuery)
        append(adapter.name)
    }

    /**
     * @return A List of column definitions that add op to a table in the DB.
     */
    fun getColumnDefinitions(): List<String> {
        val sql = "$ALTER_TABLE ${adapter.name}"
        return internalColumnDefinitions.map { "$sql ADD COLUMN $it" }
    }

    companion object {

        const val ALTER_TABLE = "ALTER TABLE "
    }
}
